### PR TITLE
Fixes unsound indexing simplification

### LINKED ISSUE #1233: [inductor] Accuracy for crossvit
https://github.com/pytorch/torchdynamo/issues/1190

~~~

import torch
import torchdynamo
from torch import tensor, device
import torch.fx as fx
from torchdynamo.testing import rand_strided
from math import inf
from torchdynamo.debug_utils import run_fwd_maybe_bwd
from torchdynamo.debug_utils import same_two_models

# Set fake tensor off as there is a known bug with it.
import functorch.compile
functorch.compile.config.use_fake_tensor = False

from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.self_blocks_2__fusion_0__norm1 = LayerNorm((256,), eps=1e-06, elementwise_affine=True)
        self.self_blocks_2__fusion_0__attn_wq = Linear(in_features=256, out_features=256, bias=True)



    def forward(self, transpose_34, cat_10):
        self_blocks_2__fusion_0__norm1 = self.self_blocks_2__fusion_0__norm1(cat_10);  cat_10 = None
        getitem_64 = self_blocks_2__fusion_0__norm1[(slice(None, None, None), slice(0, 1, None), ...)];  self_blocks_2__fusion_0__norm1 = None
        self_blocks_2__fusion_0__attn_wq = self.self_blocks_2__fusion_0__attn_wq(getitem_64);  getitem_64 = None
        reshape_40 = self_blocks_2__fusion_0__attn_wq.reshape(2, 1, 4, 64);  self_blocks_2__fusion_0__attn_wq = None
        permute_24 = reshape_40.permute(0, 2, 1, 3);  reshape_40 = None
        matmul_32 = permute_24 @ transpose_34;  permute_24 = transpose_34 = None
        return (matmul_32,)


mod = Repro().cuda()
opt_mod = torchdynamo.optimize("inductor")(mod)

mod.eval()
opt_mod.eval()

args = [((2, 4, 64, 197), (50432, 64, 1, 256), torch.float32, 'cuda', True), ((2, 197, 256), (50432, 256, 1), torch.float32, 'cuda', True)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


with torch.cuda.amp.autocast(enabled=False):
    # ref = run_fwd_maybe_bwd(mod, args)
    # res = run_fwd_maybe_bwd(opt_mod, args)
    assert same_two_models(mod, mod, args), "Eager itself failed"
    assert same_two_models(mod, opt_mod, args), "Dynamo failed"
~~~

### PR DESCRIPTION
**Pull request by ngimel**
fixes #1233
`remove_zero_terms` removed a bit too much, producing incorrect results (even indexing simplification tests were expecting incorrect results).

### CODE REVIEW

**Review by jansel — APPROVED** [review]
Nice find!

**Comment by ngimel** [comment]
Test failures unrelated

**Comment by anijain2305** [comment]
Awesome. (eagerly looking forward to tomorrow's results, it was not picked up last nightly). This gives me confidence that minifier can actually work for debugging accuracy issues as well. Let me try to extend it to Inductor, so that you don't have to further minify the repro.

**Comment by ngimel** [comment]
One thing I'd like to mention is that currently we do `loss.sum().backward()` and that masks indexing errors in the first backward op (that's the reason you needed 2 matmuls on top of faulty layer_norm here), because input to it consists of all ones. If we ran backward like `loss.backward(rand_like(loss))` that would be more sensitive
